File: apps/services/notifications/Notifications.Infrastructure/Services/TemplateResolutionService.cs
using Microsoft.Extensions.Logging;
using Notifications.Application.Interfaces;

namespace Notifications.Infrastructure.Services;

public class TemplateResolutionService : ITemplateResolutionService
{
    private readonly ITemplateRepository _templateRepo;
    private readonly ITemplateVersionRepository _versionRepo;
    private readonly ILogger<TemplateResolutionService> _logger;

    public TemplateResolutionService(ITemplateRepository templateRepo, ITemplateVersionRepository versionRepo, ILogger<TemplateResolutionService> logger)
    {
        _templateRepo = templateRepo;
        _versionRepo = versionRepo;
        _logger = logger;
    }

    public async Task<ResolvedTemplate?> ResolveAsync(Guid tenantId, string templateKey, string channel)
    {
        var tenantTemplate = await _templateRepo.FindByKeyAsync(templateKey, channel, tenantId);
        if (tenantTemplate is { Status: "active" })
        {
            var version = await _versionRepo.FindPublishedByTemplateIdAsync(tenantTemplate.Id);
            if (version != null)
            {
                _logger.LogDebug("Template resolved via tenant-specific override: {TemplateKey} {Channel} {TenantId}", templateKey, channel, tenantId);
                return new ResolvedTemplate { Template = tenantTemplate, Version = version };
            }
        }

        var globalTemplate = await _templateRepo.FindByKeyAsync(templateKey, channel, null);
        if (globalTemplate is { Status: "active" })
        {
            var version = await _versionRepo.FindPublishedByTemplateIdAsync(globalTemplate.Id);
            if (version != null)
            {
                _logger.LogDebug("Template resolved via global system template: {TemplateKey} {Channel}", templateKey, channel);
                return new ResolvedTemplate { Template = globalTemplate, Version = version };
            }
        }

        _logger.LogWarning("Template resolution failed - no active+published template found: {TemplateKey} {Channel} {TenantId}", templateKey, channel, tenantId);
        return null;
    }

    public async Task<ResolvedTemplate?> ResolveByProductAsync(Guid tenantId, string templateKey, string channel, string productType)
    {
        var globalTemplate = await _templateRepo.FindGlobalByProductKeyAsync(productType, channel, templateKey, "global");
        if (globalTemplate is { Status: "active" })
        {
            var version = await _versionRepo.FindPublishedByTemplateIdAsync(globalTemplate.Id);
            if (version != null)
            {
                _logger.LogDebug("Template resolved via product-type global template: {TemplateKey} {Channel} {ProductType}", templateKey, channel, productType);
                return new ResolvedTemplate { Template = globalTemplate, Version = version };
            }
        }

        return await ResolveAsync(tenantId, templateKey, channel);
    }
}

File: _archived/notifications-nodejs/src/services/template-resolution.service.ts
import { Template } from "../models/template.model";
import { TemplateVersion } from "../models/template-version.model";
import { TemplateRepository, TemplateVersionRepository } from "../repositories/template.repository";
import { NotificationChannel, ProductType } from "../types";
import { logger } from "../shared/logger";

export interface ResolvedTemplate {
  template: Template;
  version: TemplateVersion;
}

export class TemplateResolutionService {
  private templateRepo: TemplateRepository;
  private versionRepo: TemplateVersionRepository;

  constructor() {
    this.templateRepo = new TemplateRepository();
    this.versionRepo = new TemplateVersionRepository();
  }

  /**
   * Resolution order:
   * 1. Tenant-specific active template with a published version
   * 2. Global (tenantId=null) system template with a published version
   * 3. Fail cleanly
   */
  async resolve(
    tenantId: string,
    templateKey: string,
    channel: NotificationChannel
  ): Promise<ResolvedTemplate | null> {
    const tenantTemplate = await this.templateRepo.findByKey(templateKey, channel, tenantId);
    if (tenantTemplate && tenantTemplate.status === "active") {
      const version = await this.versionRepo.findPublishedByTemplateId(tenantTemplate.id);
      if (version) {
        logger.debug("Template resolved via tenant-specific override", {
          tenantId,
          templateKey,
          channel,
          templateId: tenantTemplate.id,
          versionId: version.id,
        });
        return { template: tenantTemplate, version };
      }
    }

    const globalTemplate = await this.templateRepo.findByKey(templateKey, channel, null);
    if (globalTemplate && globalTemplate.status === "active") {
      const version = await this.versionRepo.findPublishedByTemplateId(globalTemplate.id);
      if (version) {
        logger.debug("Template resolved via global system template", {
          tenantId,
          templateKey,
          channel,
          templateId: globalTemplate.id,
          versionId: version.id,
        });
        return { template: globalTemplate, version };
      }
    }

    logger.warn("Template resolution failed — no active+published template found", {
      tenantId,
      templateKey,
      channel,
    });
    return null;
  }

  async resolveByProduct(
    tenantId: string,
    templateKey: string,
    channel: NotificationChannel,
    productType: ProductType
  ): Promise<ResolvedTemplate | null> {
    const globalTemplate = await this.templateRepo.findGlobalByProductKey(
      productType,
      channel,
      templateKey,
      "global"
    );
    if (globalTemplate && globalTemplate.status === "active") {
      const version = await this.versionRepo.findPublishedByTemplateId(globalTemplate.id);
      if (version) {
        logger.debug("Template resolved via product-type global template", {
          tenantId,
          templateKey,
          channel,
          productType,
          templateId: globalTemplate.id,
          versionId: version.id,
        });
        return { template: globalTemplate, version };
      }
    }

    return this.resolve(tenantId, templateKey, channel);
  }
}
